### DISCUSSION TITLE
File design draft

### DISCUSSION THREAD

**Discussion by WithoutPants** [stashapp/stash#1958] [Ideas]
# Introduction

This is a design document for the introduction of storing file metadata in the database as first class objects, instead of embedding this information in media-based objects. 

This document will be updated as the design is iterated on and feedback is received.

# Background

Currently, there is a one-to-one relationship between files and media. That is, for each identified file in a stash library, there is a related scene, image or gallery object. These objects are keyed by the file checksum/oshash, and we these checksums are uniquely constrained. This means that it is not currently possible to add files with identical checksums but different paths to the stash library. Currently, these identical files are logged and ignored. 

There are legitimate reasons why a user may wish to maintain these identical files, and the system should allow management of these identical files. 

We also eventually want to decouple scenes from files, so that a single file may consist of multiple scenes.

# Goals

The immediate goal is to shift the storage of files from the `scenes`, `images` and `galleries` tables into a dedicated `files` table, and support scanning of multiple identical files into the system.

This migration should occur with minimal disruption and intervention for the user. 

Scenes, Images and Galleries should be able to be related to multiple Files. For the first iteration, this should be handled by the scan process only, and we do not allow manual manipulation of file relationships.

While not an immediate goal, an eventual goal is to completely decouple scenes, images and galleries from files so that these objects can be freely associated in a many-to-many manner.

# Design

## Decisions

* File fingerprints will be stored in a dedicated table
* Initially, file metadata (`duration`, `video_codec`, etc) will be replicated and maintained in the object tables (`scenes`, `images`, `galleries`).
* `path` field should be relative for files in zip files

### Dedicated Fingerprints table

We currently store file fingerprints as columns on the object tables - `checksum`, `oshash` and `phash`. Whenever we add support for a new fingerprint algorithm, we have to change the schema to add the column, plus add any relevant indexes. Moving this information to a dedicated table means that we can add fingerprint algorithms to the system without needing to change the schema. This could also potentially allow us to support multiple fingerprints per file - for example having phashes for multiple individual frames in a scene file.

The downside to this is that we need to join an additional table to query for object fingerprints.

### File Metadata Replication

Currently, we have a non-trivial amount of file-specific fields on objects in the graphql schema. For example, for scene:
* `checksum`
* `oshash`
* `path`
* the embedded `file` type

The underlying assumption with all of these fields is that a scene is attached to a single file. If we shift the file metadata from the `scenes` table, we have to determine which `files` row from which to get this information, since there now may be more than one. For ease of initial implementation, and minimal disruption to client code, I propose that file metadata is still used to populate the object file metadata, and to maintain our existing code to query on object file metadata in the existing tables, rather than querying on the `files` table. In this initial version, objects should only be associated with identical files, so it won't matter which file we use to set the metadata (with the exception of the `path`). In further iterations, this will be changed as the graphql interface develops and we can deprecate the file-specific fields from the object tables.

The `path` field will need to be deprecated regardless since it's current form will no longer be applicable.

The alternative is that we immediately deprecate the file-specific fields from the object tables, but this will require extensive changes to the resolver and query logic.

### Files in zip files

Currently, images in zip files are signified in the database by appending the file path in the zip to the zip path, with a `NULL` character as a separator between the two. This causes issues when querying on the path column, especially when using the `regexp` operator, which refuses to run when there is a `NULL` character in the string.

Proposal is to have a one-to-many relationship between `files` rows to represent files within a zip file. A zip files contents can then be queried by using the zip file id. The `path` field will then be the relative path within the zip file. This means that we can't have a direct unique index on path, but must use a composite index of the path and zip file id. It also means that the `path` resolver for images must query for the zip file path and append to it - presumably with a path separator between the two.

The alternative is to continue to use the current system of appending relative zip path to the zip file path with some separator. Ideally, we'd want some other character to be used as a separator since the current separator is problematic. 

## Database

Propose adding the following tables:
* `files`
* `files_fingerprints` - related to the `files` table, collects the `algorithm` and `hash` values for each object fingerprint.
* `scenes_files` - relationship table between `scenes` and `files` (many-to-many)
* `images_files` - relationship table between `images` and `files` (many-to-many)
* `galleries_files` - relationship table between `galleries` and `files` (many-to-many)

### `files` table

This table contains the superset of file-specific fields from the current `scenes`, `images` and `galleries` tables. In addition, it will include a foreign key `zip_file_id` to the `files` table. This field will be populated for image files within gallery zip files.

This table will have a unique composite index on the `path` and `zip_file_id` columns. This means that paths must be unique within zip files, and outside zip files, and means that the `path` field can be a relative path within a zip file.

## Graphql schema

The `scene` type, currently has a single mandatory `path` string. It also has an unrelated `paths` field for `ScenePathsType` which provides URL paths for the client. Finally, it provides a `file` field for `SceneFileType`. 

Propose to deprecate the `file` and `path` fields. These fields will provide the file details from the _first_ related file to the scene object. Propose to add new `files` field to `Scene`. This will be a list of `SceneFileType` objects. `SceneFileType` should be augmented to provide `path`.

Future iterations should look at deprecating `checksum` and `oshash` and moving these into the `SceneFileType` type, probably as a separate `FingerprintType` type to follow the database schema changes.

Similar changes will be introduced for `Image` and `Gallery`.

Note that the existing `GalleryFilesType` is currently unused and can be repurposed.

### Generating and viewing media

To simplify implementation, we will assume that the first referenced file for an object will be used for generation and presentation. In further iterations, we will want to consider allowing the user to choose the file to view, especially where other files may be encoded in different resolutions, for example.

### File Info tab

The `Path` field should present all paths for the object. For the initial implementation, the file details from the first referenced file will be used for the other fields.

We may wish to consider adding functionality to delete an object's file if there are more than one.

### Auto Tagger

The auto tagger code will be adapted so that it will auto-tag based on all of the object's file paths.

### Import/Export

Propose changing the mappings schema to deprecate `path` (but maintain import compatibility) and add `paths`. In future, the import/export system should be refactored further to not be as tightly coupled to paths and fingerprints.

### Passing paths to scrapers and plugins

For ease of implementation, the first path will be passed to scrapers and plugins where applicable. This should be revisited in future though.

### Querying

We should provide a `files_count` filter criterion so that it is easy to find scenes with duplicate files.

### Deleting objects

If the `delete_file` flag is true, then _all_ files related to the object should be deleted.

## Scanning and Cleaning

We already have a `file` package which contains a `Scanner` type that consolidates the basic file scanning logic. This will be extended further to create and potentially destroy `File` objects in the database. The client scanning code in `scene`, `image` and `gallery` will manage the relationships between files and objects.

The clean task logic should be formalised in a similar Scanner-like type (or add to the existing `Scanner` types). The only major behavioural change is that an object should only be destroyed if _all_ related files are now missing.

## Unresolved questions

* how should the filename parser be adjusted to allow for scenes with more than one path?

# Relevant issues

* #727
* #1088
* #837

**Comment by SmallCoccinelle** [comment]
I like the idea of not changing any UX but only the backend structure in a first iteration (to the extent this is possible). It allows one to take that step, then iterate a bit, before opening the "flood gates".

I also like the idea of handling GraphQL fields where possible, and return `null` for fields where this can't happen due to the underlying constraint change. Multiplicity of files will have some ripple-effects through the system, and if we can keep some of that stable, it's going to minimize changes. Once things settle, you can then decide what to do.

I've been looking a bit into duplicate handling, and everything proposed here would make that handling way simpler. Curation for a particular scene is likely to live on table `scenes_files`, for instance. So my educated guess is that this design will solve a lot of problems / requests people have down the road. Our issue tracker is likely only looking at the top of the iceberg.

**Reply by philpw99** [reply]
Love the idea of multiple files for one scene, but not really for the retiring of existing File/OShash/checksum.
Because the most common situation is one file to one scene. The new mechanism solves the problem of additional files for the same scene, but out of 100 scenes, maybe only 1 or 2 has duplicate files or files in different size/resolution. Of course there are situations that people will want to keep the duplicates, but in most cases they don't. So it's like a big overhaul to solve a small problem.
I propose to just directly link to a "Files" table. The structure of Files is like below:
| Scene ID | Scene Start Time | Scene End Time | Main File | File location | File Type | File Length | OShash | Checksum |
| ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| 11 | 0:00 | 10:00 | Yes | D:\Video\myvideo.mp4 | Video | 10:00 | A1B2C3D4 | E5F6123456 |
| 11 | 0:00 | 10:00 | No | D:\Video\other\myvideo2.mp4 | Video | 10:00 | A1B2C3D4 | 123456ABCD |

This way, the "File" in the scenes table is always the main file, and there should be only one "main file" for each scene. If you start to delete duplicate files, you can delete all others, but you cannot delete the ones that's a "main file".
You don't need a multi-to-multi relationship with this setup. Scene 2 with the same file can simple have another entry in the Files table. This will lower the complexity of programming, and lower the risk of deleting important files.
This will also solve the problem of "which file to play". Sometimes a user actually want to play a lower resolution video. So by assigning the main file, stash will work better in their cellphones/tablets.

For image files, the start time and end time are always null. For image files in a zip file, the location should be the full path and file name of that zip file.
For different type of file finger prints. I think using a field: Fingerprints is good enough. The content is this:
```
MD5: 12345379ADDA
SHA-1: 1BCDFA4AASS
......
```
The most important hashes should have their own field, but for all others, just put them in the same field. We can definitely search them if necessary.

**Reply by SmallCoccinelle** [reply]
For the last comment: In Postgres, my solution would be to have all fingerprint types in a `jsonb` column and then have an index on `fingerprint->oshash` etc. If you really want it as a column, that would just be a VIEW.

In sqlite, I'm not so sure however. I don't know the extent of its JSON handling power.

**Reply by WithoutPants** [reply]
> Love the idea of multiple files for one scene, but not really for the retiring of existing File/OShash/checksum. Because the most common situation is one file to one scene. The new mechanism solves the problem of additional files for the same scene, but out of 100 scenes, maybe only 1 or 2 has duplicate files or files in different size/resolution. Of course there are situations that people will want to keep the duplicates, but in most cases they don't. So it's like a big overhaul to solve a small problem. I propose to just directly link to a "Files" table. The structure of Files is like below:
> 
> Scene ID	Scene Start Time	Scene End Time	Main File	File location	File Type	File Length	OShash	Checksum
> 11	0:00	10:00	Yes	D:\Video\myvideo.mp4	Video	10:00	A1B2C3D4	E5F6123456
> 11	0:00	10:00	No	D:\Video\other\myvideo2.mp4	Video	10:00	A1B2C3D4	123456ABCD

> You don't need a multi-to-multi relationship with this setup. Scene 2 with the same file can simple have another entry in the Files table. This will lower the complexity of programming, and lower the risk of deleting important files.

Consider the eventual use case where we have a single movie file which will logically consist of multiple scenes. We now have to duplicate _all_ of the file information just for the purposes of linking to the scene ID with different start/end times. 

I don't think polluting the file concept with scene/image/gallery links is going to simplify the programming. As stated above, duplicating the entire file data for the purposes of linking another scene seems horribly inefficient to me. We can no longer uniquely identify a file by its path. If a files details change - hash, mod time - or if it is renamed, we now have to iterate through multiple file entries.

Scene ID, scene start and end time and main file fields do not belong in the files table. The purpose of the files table is to represent files in the file system _only_. This allows us to operate on these objects in isolation without concerning ourselves with the relationship to other objects.

Keep in mind that this overhaul is not only for duplicate file handling, it is just what I'm looking to solve specifically in this iteration.

The concept of a main or primary file is a possible iteration and one I've played around with in prototyping. It's biggest utility is in determining what file to play by default. I don't think it's worth adding until we have performed the initial implementation and it's utility emerges from actual use cases.

> For different type of file finger prints. I think using a field: Fingerprints is good enough. The content is this:
> 
> ```
> MD5: 12345379ADDA
> SHA-1: 1BCDFA4AASS
> ......
> ```
> 
> The most important hashes should have their own field, but for all others, just put them in the same field. We can definitely search them if necessary.

The entire purpose of putting fingerprints into their own table is to index them. This would not be suitable.

**Comment by philpw99** [comment]
The "Start Time" and "End Time" is to handle the one file -> many scenes problem.  A video file could be like 2 hours long and contains 3 scenes.
And Yes, now I see your purpose of using the File table. When doing a file scanning and files have changed, it's better to have only one entry for each file.
I was just trying to propose a simplified implementation of files. One table for all the files registries. For me it just seemed too complicated to create 3 many-to-many tables plus a Files table to do it. Maybe instead of three many-to-many, we can do it with just one many-to-many ?

Table: items_files
| Item ID | Item Type | File ID |
| ----- | ----- | ----- |
| 11 | Scene | 1001 |
| 101 | Image | 1002 |

For fingerprints, completely agree with you. This is like my Barcode implement,  eventually I have to create a table just for all the barcodes.

**Comment by WithoutPants** [comment]
I've been doing a lot of prototyping recently to try to push this functionality along. I thought I'd share what I hope to make the eventual data model:

![image](https://user-images.githubusercontent.com/53250216/162891830-c7185cdd-4834-4515-8f51-994ec75ba907.png)

The addition of the `Folder` class allows for querying by path without having to store the full path of every file in the system, which is likely to be expensive and redundant. 

`last_scanned` can be used while scanning to detect missing files. The field is updated as it is scanned (and any `missing_since` value is nulled), then we can mark anything with `last_scanned` < `start scan time` as missing. Clean is simply a matter of deleting anything where `missing_since` is not null.

Scenes, Images and Galleries become logical entities. Images and galleries have a direct relationship to files. Typically, this will be 1, but may be more if there are duplicate files. Scenes and movies are associated to files via `FileRange` objects. This enables us to have scenes and movies that are split across multiple files, or scenes that are a sub-section of a file. In many cases the start/end values can be null to indicate a one-to-one relationship from scene/movie to file.

Likewise with fingerprints, they can refer to a single fingerprint for an entire scene/movie, or a sub-section of a file. The relationship to fingerprints is duplicated from the file relationship so that scenes/movies can have fingerprints independently of files. This allows a file to be deleted, but the movie/scene to stay and then reassociate with a file based on the fingerprints.

Edit: the following is a sequence diagram for the scanning algorithm

![image](https://user-images.githubusercontent.com/53250216/163542342-071e1bc8-0ea9-45d9-a2d4-4359c596f644.png)


(plantUML code for reference)
```
@startuml
class File {
  id: int32
  basename: string
  
  size: int64
  mod_time: Time
  
  last_scanned: Time
  missing_since: Time

  created_at: Time
  updated_at: Time
}

class Folder {
  id: int32
  path: string
  
  mod_time: Time
  
  last_scanned: Time
  missing_since: Time

  created_at: Time
  updated_at: Time
}

class Fingerprint {
  type: string
  value: any
}

class FileRange {
  start: float
  end: float
}

class FingerprintRange {
  start: float
  end: float
}

class Image

class Scene

class Movie

class Gallery

Folder "1" <- File : parent_folder
File -> "0..1" File: zip_file
File *-> "0..n" Fingerprint: fingerprints

Gallery --> "0..n" Image: images
Gallery -> "0..n" Scene: scenes
Gallery --> "0..n" File: files

Image --> "0..n" File: files
FileRange --> "1" File: file

Scene *--> "0..n" FileRange: file_ranges

Movie *--> "0..n" FileRange: file_ranges
Scene "0..n" <- Movie: scenes

Scene *--> "0..n" FingerprintRange: fingerprints
Movie *--> "0..n" FingerprintRange: fingerprints
FingerprintRange --> "1" Fingerprint: fingerprint
@enduml
```

PlantUML for sequence diagram:
```
 @startuml
start
repeat
  :read directory entry;
  if (is directory?) then (directory)
    :get folder by path;
    if (exists?) then (exists)
      :update modtime\nand scan time\nclear missing flag;
    else (new)
      :create new\nfolder;
    endif
  else (file)
    :get file by path;
    if (exists?) then (exists)
      :update scan time\nclear missing flag;
      if (mod time changed?) then (changed)
        :recalculate details\nand fingerprints;
        :fire handlers;
      else (unchanged)
      endif
    else (new)
      :calculate fingerprints;
      :find existing files\nwith fingerprint;
      :check if each file\nexists in filesystem;
      if (missing files?) then (is 1)
        :assume rename\nupdate path of\nexisting file;
      else (0 or > 1)
        :create file;
        :fire handlers;
      endif
    endif
  endif
repeat while (more entries?) is (yes)
->no;
:mark files not\nseen as missing;
stop
@enduml
```

**Comment by laurus-lx** [comment]
Is this proposal still actively developed/considered?
IMHO It's really well thought out and would address a lot of missing functionality / use cases

Feature requests that would be directly related (not already covered in this thread):
https://github.com/stashapp/stash/issues/4642
https://github.com/stashapp/stash/issues/2511
https://github.com/stashapp/stash/issues/3530

**Reply by WithoutPants** [reply]
It serves as guidance for #3530 and others, yes. It is nonetheless a very complex change, but part of the roadmap.
